inference-engine/src/gna_plugin/optimizer/gna_pass_manager.cpp

```diff
@@ -185,19 +187,92 @@ static std::vector<CNNLayerPtr> getCandidatesForIdentityInsertion(const CNNLayer
                 prevLayers.push_back(CNNNetPrevLayer(l, i));
             }
         }
+    } else if (LayerInfo(l).isOutput() && (LayerInfo(l).isPooling() || LayerInfo(l).isNonFunctional()))  {
+        if (LayerInfo(l).isPooling()) {
+            // in GNA pooling get executed on 2 bytes data, that is why preceding activation required
+            // in gna - primitive fusing we however fuse different order conv->pool->act
+            prevLayers.push_back(l);
+        }
+        if (LayerInfo(l).isNonFunctional()) {
+            // for output layers even if it is non functional it might have limitations
+            // gna cannot have 32bits precision for pooling layer output
+            auto prev = PrevFunctionalLayer(l, 0);
+            if (LayerInfo(prev).isPooling()) {
+                prevLayers.push_back(prev);
+            }
+        }
     } else {
         // not eltwise or concat
         // other layers has 1 inputs - situation is easier
-        // ex. activation or pooling - no need to insert identity activation.
+        // ex. activation  - no need to insert identity activation.
         if (LayerInfo(l).isNonFunctional() || LayerInfo(l).has32BInput())
             return prevLayers;
 
-        auto prevLayer = PrevFunctionalLayer(l, 0);
+        gnalog() << "CNNNetPrevLayer skip non functional from :: " << l->name;
+        auto isFunctional = [](CNNLayerPtr ptr) {
+            return !LayerInfo(ptr).isNonFunctional();
+        };
+        auto isNonFunctional = [](CNNLayerPtr ptr) {
+            return LayerInfo(ptr).isNonFunctional();
+        };
+
+        auto prevLayersReached = CNNNetGetPrevLayersSkip(l, isFunctional);
+
+        prevLayersReached.erase(std::remove_if(prevLayersReached.begin(),
+            prevLayersReached.end(),
+            [] (const std::pair<CNNLayerPtr, int> & candidate) {
+                return LayerInfo(candidate.first).isLink();
+            }), prevLayersReached.end());
+        if (prevLayersReached.size() != 1) {
+            std::stringstream layers;
+            for (auto && prevLayer : prevLayersReached) {
+                layers << prevLayer.first->name;
+                layers << ", ";
+            }
+            THROW_GNA_LAYER_EXCEPTION(l) << "unsupported case: connected to "
+                << (prevLayersReached.empty() ? "zero" : "multiple") << " outputs : " << layers.str();
+        }
+        auto prevLayer = prevLayersReached.front().first;
+        auto outDataIdx = prevLayersReached.front().second;
+        gnalog() << ", reached " << prevLayer->name << " at " << outDataIdx << std::endl;
 
         if (!LayerInfo(prevLayer).has32BOutput())
             return prevLayers;
 
-        prevLayers.push_back(CNNNetPrevLayer(l, 0));
+        // identity might be already attached to same  layer output but hidden by non functional childs
```

Comment:
please remove that - we transformed this code into dedicated pass : fuse multipleIdentities

Reply:
Ok, will do